exp/etcdrestore/controllers/etcdmachinesnapshot_controller.go

```diff
@@ -60,18 +67,224 @@ func (r *EtcdMachineSnapshotReconciler) SetupWithManager(_ context.Context, mgr
 //+kubebuilder:rbac:groups=turtles-capi.cattle.io,resources=etcdmachinesnapshots/finalizers,verbs=update
 
 // Reconcile reconciles the EtcdMachineSnapshot object.
-func (r *EtcdMachineSnapshotReconciler) Reconcile(_ context.Context, _ ctrl.Request) (ctrl.Result, error) {
-	return ctrl.Result{}, nil
+func (r *EtcdMachineSnapshotReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := log.FromContext(ctx)
+
+	etcdMachineSnapshot := &snapshotrestorev1.EtcdMachineSnapshot{ObjectMeta: metav1.ObjectMeta{
+		Name:      req.Name,
+		Namespace: req.Namespace,
+	}}
+	if err := r.Client.Get(ctx, req.NamespacedName, etcdMachineSnapshot); apierrors.IsNotFound(err) {
+		// Object not found, return. Created objects are automatically garbage collected.
+		return ctrl.Result{}, nil
+	} else if err != nil {
+		log.Error(err, fmt.Sprintf("Unable to get etcdMachineSnapshot resource: %s", req.String()))
+		return ctrl.Result{}, err
+	}
+
+	// Handle deleted etcdMachineSnapshot
+	if !etcdMachineSnapshot.ObjectMeta.DeletionTimestamp.IsZero() {
+		return r.reconcileDelete(ctx, etcdMachineSnapshot)
+	}
+
+	// Ensure the finalizer is present
+	if !controllerutil.ContainsFinalizer(etcdMachineSnapshot, snapshotrestorev1.ETCDMachineSnapshotFinalizer) {
+		controllerutil.AddFinalizer(etcdMachineSnapshot, snapshotrestorev1.ETCDMachineSnapshotFinalizer)
+		if err := r.Client.Update(ctx, etcdMachineSnapshot); err != nil {
+			log.Error(err, "Failed to add finalizer to EtcdMachineSnapshot")
+			return ctrl.Result{}, err
+		}
+	}
+
+	return r.reconcileNormal(ctx, etcdMachineSnapshot)
 }
 
 func (r *EtcdMachineSnapshotReconciler) reconcileNormal(
-	_ context.Context, _ *snapshotrestorev1.EtcdMachineSnapshot,
-) (_ ctrl.Result, err error) {
+	ctx context.Context, etcdMachineSnapshot *snapshotrestorev1.EtcdMachineSnapshot,
+) (ctrl.Result, error) {
+	log := log.FromContext(ctx)
+
+	// Handle different phases of the etcdmachinesnapshot creation process
+	switch etcdMachineSnapshot.Status.Phase {
+	case "":
+		// Initial phase, set to Pending
+		etcdMachineSnapshot.Status.Phase = snapshotrestorev1.ETCDSnapshotPhasePending
+		if err := r.Client.Status().Update(ctx, etcdMachineSnapshot); err != nil {
+			log.Error(err, "Failed to update EtcdMachineSnapshot status to Pending")
+			return ctrl.Result{}, err
+		}
+	case snapshotrestorev1.ETCDSnapshotPhasePending:
+		// Transition to Running
+		etcdMachineSnapshot.Status.Phase = snapshotrestorev1.ETCDSnapshotPhaseRunning
+		if err := r.Client.Status().Update(ctx, etcdMachineSnapshot); err != nil {
```

Comment:
It should be possible to use Patch here as well. Best upstream (CAPI) practice to writing controllers uses `defer` to perform patch on modified resources at the end of the full reconcile cycle.

Reply:
We can only set the phase here and return, `Reconcile()` will patch the object and get to the next phase in the next reconciliation cycle. This way we will also signal to the users the exact status of the snapshot.